### DISCUSSION TITLE
Wrap Single .so file

### DISCUSSION THREAD

**Discussion by alcubierre-drive** [ptitSeb/box64#207] [Q&A]
I was wondering whether it would be possible to use the functionality of this project to wrap a single library to provide an aarch64 version of a x86_64 .so file.

**Comment by ptitSeb** [comment]
It maybe be possible yes. It would depend on the lib. C lib might work. C++ or other language probably not.

**Reply by alcubierre-drive** [reply]
I'm particularly interested in wrapping libwidevinecdm.so as this library doesn't exist on aarch64... It should be C, not C++

**Reply by ptitSeb** [reply]
a `nm -D libwidevinecdm.so` would give some insight on how many symbols are present.
The other difficulty would be to find the signature of the functions (maybe in chrome sources?).

**Reply by alcubierre-drive** [reply]
U abort
                 U bcmp
                 U catclose
                 U catgets
                 U catopen
                 U clock_gettime
0000000000446890 T CreateCdmInstance
                 U __ctype_get_mb_cur_max
                 U __cxa_atexit
                 w __cxa_finalize
0000000000446850 T DeinitializeCdmModule
                 U dlclose
                 U dlsym
                 U __errno_location
                 U exp
                 U fclose
                 U fcntl
                 U fflush
                 U fgets
                 U fopen64
                 U fprintf
                 U fputc
                 U freelocale
                 U fwrite
                 U getauxval
                 U getc
0000000000446a20 T GetCdmVersion
                 U getenv
                 U getpid
                 U gettimeofday
                 w __gmon_start__
0000000000446820 T InitializeCdmModule_4
                 U __isnan
                 U __isoc99_sscanf
                 U __isoc99_vsscanf
                 w _ITM_deregisterTMCloneTable
                 w _ITM_registerTMCloneTable
                 U __libc_calloc
                 U __libc_free
                 U __libc_malloc
                 U __libc_memalign
                 U __libc_realloc
                 U localtime_r
                 U longjmp
                 U madvise
                 U mbrtowc
                 U mbsnrtowcs
                 U mbsrtowcs
                 U mbtowc
                 U memcmp
                 U memcpy
                 U __memcpy_chk
                 U memmove
                 U memset
                 U mmap64
                 U munmap
                 U nanosleep
                 U newlocale
                 U open64
                 w OPENSSL_memory_alloc
                 w OPENSSL_memory_free
                 w OPENSSL_memory_get_size
                 U perror
                 U prctl
                 U pthread_attr_destroy
                 U pthread_attr_init
                 U pthread_attr_setstacksize
                 U pthread_cond_broadcast
                 U pthread_cond_destroy
                 U pthread_cond_init
                 U pthread_cond_signal
                 U pthread_cond_wait
                 U pthread_create
                 U pthread_getspecific
                 U pthread_join
                 U pthread_key_create
                 U pthread_mutexattr_destroy
                 U pthread_mutexattr_init
                 U pthread_mutex_destroy
                 U pthread_mutex_init
                 U pthread_mutex_lock
                 U pthread_mutex_trylock
                 U pthread_mutex_unlock
                 U pthread_once
                 U pthread_rwlock_destroy
                 U pthread_rwlock_init
                 U pthread_rwlock_rdlock
                 U pthread_rwlock_unlock
                 U pthread_rwlock_wrlock
                 U pthread_self
                 U pthread_setspecific
                 U qsort
                 U rand
                 U read
                 U __register_atfork
                 U __sched_cpucount
                 U sched_getaffinity
                 w sdallocx
                 U sem_destroy
                 U sem_init
                 U sem_post
                 U sem_trywait
                 U sem_wait
                 U _setjmp
                 U snprintf
                 U __stack_chk_fail
                 U stderr
                 U stdin
                 U stdout
                 U strchr
                 U strcmp
                 U strerror_r
                 U strftime_l
                 U strlen
                 U strtod_l
                 U strtof_l
                 U strtol
                 U strtold_l
                 U strtoll_l
                 U strtoul
                 U strtoull_l
                 U syscall
                 U sysconf
                 U __tls_get_addr
                 U ungetc
                 U _Unwind_DeleteException
                 U _Unwind_GetIP
                 U _Unwind_GetLanguageSpecificData
                 U _Unwind_GetRegionStart
                 U _Unwind_RaiseException
                 U _Unwind_Resume
                 U _Unwind_SetGR
                 U _Unwind_SetIP
                 U uselocale
                 U usleep
                 U vasprintf
0000000000446a30 T VerifyCdmHost_0
                 U vfprintf
                 U vsnprintf
                 U __vsnprintf_chk
                 U wcrtomb
                 U wcslen
                 U wcsnrtombs
                 U wmemcpy
                 U wmemmove
                 U wmemset

**Comment by alcubierre-drive** [comment]
The nm -D output is easy to get, for the signatures I'd have to dig through the chromium sources more thoroughly...

**Reply by ptitSeb** [reply]
It seems there is just 5 actual functions:
```
0000000000446890 T CreateCdmInstance
0000000000446850 T DeinitializeCdmModule
0000000000446a20 T GetCdmVersion
0000000000446820 T InitializeCdmModule_4
0000000000446a30 T VerifyCdmHost_0
```
That probably means that one of the function (like `CreateCdmInstance`) return a pointer to a structure with a gazillion of functions... Wich is a pain to wrap.

**Reply by alcubierre-drive** [reply]
So it might be easier to set up a 32bit chroot on an aarch64 device to run the arm7l version of chromium?

**Reply by ptitSeb** [reply]
I don't know. We need the signature of those 5 functions to be sure.

**Reply by alcubierre-drive** [reply]
From quick research I cannot find the signature in the chromium sources... I'll update you in case there are any news. Thank you for the quick responses!

**Reply by alcubierre-drive** [reply]
https://searchfox.org/mozilla-central/source/dom/media/gmp/widevine-adapter/content_decryption_module.h

**Reply by alcubierre-drive** [reply]
I could actually find it!

**Reply by ptitSeb** [reply]
Seems to quite heavy on the c++ side. I need to study that later, but it seems heavy / difficult yes.

**Reply by alcubierre-drive** [reply]
```
CDM_API void INITIALIZE_CDM_MODULE();

CDM_API void DeinitializeCdmModule();

// Returns a pointer to the requested CDM Host interface upon success.
// Returns NULL if the requested CDM Host interface is not supported.
// The caller should cast the returned pointer to the type matching
// |host_interface_version|.
typedef void* (*GetCdmHostFunc)(int host_interface_version, void* user_data);

// Returns a pointer to the requested CDM upon success.
// Returns NULL if an error occurs or the requested |cdm_interface_version| or
// |key_system| is not supported or another error occurs.
// The caller should cast the returned pointer to the type matching
// |cdm_interface_version|.
// Caller retains ownership of arguments and must call Destroy() on the returned
// object.
CDM_API void* CreateCdmInstance(int cdm_interface_version,
                                const char* key_system,
                                uint32_t key_system_size,
                                GetCdmHostFunc get_cdm_host_func,
                                void* user_data);

CDM_API const char* GetCdmVersion();
```

**Reply by alcubierre-drive** [reply]
These should be the corresponding c functions/types

**Reply by ptitSeb** [reply]
Ses, but the issue is with this callback: `GetCdmHostFunc` that will basicaly just return a `void*` that need to be casted to the relevant signature function, like a `glXGetProcAddress(...)`, and this is were things gets messy.
